### PR TITLE
use reference clones for submodules

### PR DESCRIPTION
**Pull request by jonahbull**
Add support for using reference repositories for submodules in addition to regular repositories.

First, hugops 🤗 for those dealing with [the current incident](https://www.buildkitestatus.com/incidents/07260v5fffm8).

This feature is listed as being supported [in the documentation](https://buildkite.com/docs/agent/v3#experimental-features-git-mirrors) (emphasis added)
> Checkouts reference the git mirror using git clone --reference, **as do submodules**.

but as far as I can tell support for submodules was removed from [the original PR implementing this feature](https://github.com/buildkite/agent/pull/936) in [`93b80d2` (#936)](https://github.com/buildkite/agent/pull/936/commits/93b80d25187a9f03901a0976ca5590efecc80054).

I'm not sure why reference repo support for submodules was removed, but it'd be very valuable to have for several of our use cases. The changes in this PR are largely based on the initial implementation that was removed, with a few tweaks. I rearranged the loop over `submoduleRepos` and added some logic to ignore errors from git if the remote for the submodule already exists on the repository. That bit's not very clever, so as always I welcome any suggestions for improvement or discussion on whether this is a feature that y'all want to support.

### CODE REVIEW

**Comment by triarius** [comment]
Hi @jonahbull, thanks for your support! We've decided we do want to support this feature, so I'll have a look at this PR and run CI on it. I'll get back to you with any feedback shortly.

File: bootstrap/bootstrap.go
Comment (triarius) [reply bootstrap/bootstrap.go:1423]: Hmmm, I looked at the `shell.GetExitCode(err)` method, and it does not inspire a log of confidence that it will work on Windows. But I've made #1938 to make it more cross-platform.

It should be fine to merge this PR with my suggestion after that's merged, but I think it would be safest to rebase.

**Review by triarius — COMMENTED** [review]
Thanks for this @jonahbull. Generally it's looking good, and I tested it works myself. But, I'm going to ask that you change the error handling.

Unfortunately, that depends on us merging another PR first, so I'm going to beg for a little patience. Feel free to @ me to approve it after accepting the suggestions and #1938 is merged.

File: bootstrap/bootstrap.go
Comment (jonahbull) [reply bootstrap/bootstrap.go:1423]: Thanks for the quick feedback @triarius, your suggestion is much cleaner 👍🏻 

Would you prefer that I used the magic "Commit suggestion" button or apply your suggestion to my local branch, rebase and force push?

FWIW when I test it on Windows it works as expected with or without your changes from #1938 but I don't have any objections to waiting on that getting merged.

Comment (triarius) [reply bootstrap/bootstrap.go:1423]: > Would you prefer that I used the magic "Commit suggestion" button or apply your suggestion to my local branch, rebase and force push?

That's fine by me. I will run CI on this again before merging.

> FWIW when I test it on Windows it works as expected with or without your changes from https://github.com/buildkite/agent/pull/1938 but I don't have any objections to waiting on that getting merged.

Oh nice, #1938 is merged now in any case.

**Comment by jonahbull** [comment]
@triarius Thanks, I applied your suggestion and rebased so this should be ready for another round of testing.

**Review by triarius — APPROVED** [review]
Thanks! LGTM

**Comment by triarius** [comment]
There are two things that need to be solved for this  work to make if back in to the main branch

1. Each submodule repo would append the arguments `submodule update --init --recursive --force` to the git command. This means that if there were 2 submodule, the command `git submodule update --init --recursive --force submodule update --init --recursive --force` would be run from https://github.com/jonahbull/agent/blob/8062f98497266749c1ae68ae2fd44dabc80a59c6/bootstrap/bootstrap.go#L1439
2. The refrence repos for submodules don't seem to be be created on the first run. This is in contrast to the when using a reference repo for the top level.

**Pull request merged**
